gluonnlp/data/stream.py

```diff
@@ -0,0 +1,313 @@
+        - 'random': RandomSampler
+    file_sampler : str, {'sequential', 'random'}, defaults to 'random'
+        The sampler used to sample a file.
+
+        - 'sequential': SequentialSampler
+        - 'random': RandomSampler
+    """
+    def __init__(self, file_pattern, encoding='utf8', skip_empty=True,
+                 sample_splitter=lambda s: s.splitlines(), tokenizer=lambda s: s.split(),
+                 bos=None, eos=None, sampler='random', file_sampler='random'):
+        self._file_pattern = file_pattern
+        self._encoding = encoding
+        self._skip_empty = skip_empty
+        self._sample_splitter = sample_splitter
+        self._tokenizer = tokenizer
+        self._bos = bos
+        self._eos = eos
+        self._sampler = sampler
+        self._file_sampler = file_sampler
+        super(LanguageModelStream, self).__init__(self._file_pattern, flatten=True,
+                                                  encoding=self._encoding,
+                                                  skip_empty=self._skip_empty,
+                                                  sample_splitter=self._sample_splitter,
+                                                  tokenizer=self._tokenizer, bos=self._bos,
+                                                  eos=self._eos, sampler=self._sampler,
+                                                  file_sampler=self._file_sampler)
+
+    def bptt_batchify(self, vocab, seq_len, batch_size, last_batch='keep'):
+        """The corpus is transformed into batches of numericalized samples, in the way that the
+        recurrent states from last batch connects with the current batch for each sample.
+
+        Each sample is of shape `(seq_len, batch_size)`.
+
+        Parameters
+        ----------
+        vocab : gluonnlp.Vocab
+            The vocabulary to use for numericalizing the dataset. Each token will be mapped to the
+            index according to the vocabulary.
+        seq_len : int
+            The length of each of the samples for truncated back-propagation-through-time (TBPTT).
+        batch_size : int
+            The number of samples in each batch.
+        last_batch : {'keep', 'discard'}
+            How to handle the last batch if the remaining length is less than `seq_len`.
+
+            - keep: A batch with less samples than previous batches is returned.
+            - discard: The last batch is discarded if it's smaller than `(seq_len, batch_size)`.
+        """
+        corpus = CorpusStream(self._file_pattern, flatten=False, encoding=self._encoding,
+                              skip_empty=self._skip_empty, sample_splitter=self._sample_splitter,
+                              tokenizer=self._tokenizer, bos=self._bos, eos=self._eos,
+                              sampler=self._sampler, file_sampler=self._file_sampler)
+        return _LanguageModelBPTTStream(corpus, vocab, seq_len, batch_size, last_batch=last_batch)
+
+class _LanguageModelBPTTStream(DataStream):
```

Comment:
We probably should provide character level vocab and language model stream as well.

Reply:
Yes. We can extend this class in later PRs.

Reply:
Yes, let's add the character vocab to support the implementation of the whole model zoo.

Reply:
we already have a flavor of character vocab in word embedding training. also, the current class already can support character level data by switching the sample splitter.